homeassistant/components/elmax/common.py

```diff
@@ -13,163 +14,226 @@
+            event_signal = f"{SIGNAL_PANEL_UPDATE}-{self.panel_entry.hash}-{k}"
+            async_dispatcher_send(self.hass, event_signal, ep_status)
+
+        self.async_set_updated_data(status)
+
+    def _register_push_notification_handler(self):
+        ws_ep = (
+            f"{'wss' if self.http_client.base_url.scheme == 'https' else 'ws'}"
+            f"://{self.http_client.base_url.host}"
+            f":{self.http_client.base_url.port}"
+            f"{self.http_client.base_url.path}/push"
+        )
+        self._push_notification_handler = PushNotificationHandler(
+            endpoint=str(ws_ep),
+            http_client=self.http_client,
+            ssl_context=self.http_client.ssl_context,
+        )
+        self._push_notification_handler.register_push_notification_handler(
+            self._push_handler
+        )
+        self._push_notification_handler.start(loop=self.hass.loop)
+
+    async def _push_handler(self, status: PanelStatus) -> None:
+        self._fire_data_update(status)
+
+    async def async_shutdown(self) -> None:
+        """Cancel any scheduled call, and ignore new runs."""
+        if self._push_notification_handler is not None:
+            self._push_notification_handler.unregister_push_notification_handler(
+                self._push_handler
+            )
+            self._push_notification_handler.stop()
+        self._push_notification_handler = None
+        return await super().async_shutdown()
+
 
 class ElmaxEntity(CoordinatorEntity[ElmaxCoordinator]):
     """Wrapper for Elmax entities."""
 
+    _last_state: DeviceEndpoint = None
+
     def __init__(
         self,
-        panel: PanelEntry,
         elmax_device: DeviceEndpoint,
         panel_version: str,
         coordinator: ElmaxCoordinator,
     ) -> None:
         """Construct the object."""
         super().__init__(coordinator=coordinator)
-        self._panel = panel
         self._device = elmax_device
         self._attr_unique_id = elmax_device.endpoint_id
         self._attr_name = elmax_device.name
         self._attr_device_info = DeviceInfo(
-            identifiers={(DOMAIN, panel.hash)},
-            name=panel.get_name_by_user(
+            identifiers={(DOMAIN, coordinator.panel_entry.hash)},
+            name=coordinator.panel_entry.get_name_by_user(
                 coordinator.http_client.get_authenticated_username()
             ),
             manufacturer="Elmax",
             model=panel_version,
             sw_version=panel_version,
         )
 
+    async def async_added_to_hass(self) -> None:
+        """Register push notifications callbacks if available."""
+        self.async_on_remove(
+            async_dispatcher_connect(
+                self.hass,
+                f"{SIGNAL_PANEL_UPDATE}-{self.coordinator.panel_entry.hash}-{self._device.endpoint_id}",
+                self._handle_update,
+            )
+        )
+        return await super().async_added_to_hass()
```

Comment:
Right so if I understand correctly, there is now a whole dispatcher construction to update the entities when they are in push state. Do I have this correct?